dashboard/src/actions/toastActions.js

```diff
@@ -10,7 +10,7 @@ export const showSessionExpired = () => async (dispatch) => {
     message: "Please login to continue",
   };
   dispatch(showToast(toast.variant, toast.title, toast.message));
-  dispatch(logout());
+  window.location.href = APP_ROUTES.AUTH
```

Comment:
You _don't_ want to do the logout action?  (It would delete the cookies and call `keycloak.logout()`...these are things which aren't important or which we want to _avoid_ in this case?)

Reply:
What I was thinking is just because user received a 401 from the _Pbench server_ API doesn't mean we should log the user out of his/her SSO session, right? I think the right course of action would be to redirect the user back to the Auth page and remove all the cookies that the dashboard is using. I am not even sure whether we should provide a logout option for the user to logout of the SSO session but that's a different discussion.

Reply:
Well, I guess I have to ask...what's an "SSO session"?  That is, if a user logs out of _one OIDC client_, does that invalidate their logins with **all** OIDC clients that the user is currently using?  I.e., is there a single SSO session shared among all the clients?...can one client really end the session of another client??

It would make more sense (to me) if, say, Pbench had its own "session" which was separate from, say, Workday's session, which in turn was separate from The Source's session, etc.  If that's the case, then it's not unreasonable for the user to end the Pbench session by logging out, and doing so should leave the other sessions unaffected.

As for session _expiration_, I was assuming that that would, roughly speaking, never happen, because the Keycloak object in the Dashboard would keep renewing the token (and, thereby, I boldly assume, the session expiration).  (Am I conflating two independent things, here?)  I suppose an idle session in an open browser (perhaps in an inactive tab) would _eventually_ expire, at which point we _should_ kick the user back to the login page (assuming that the Keycloak object doesn't already handle this, too).

> I am not even sure whether we should provide a logout option for the user to logout of the SSO session

Each of [Workday](https://wd5.myworkday.com/redhat/d/home.htmld) and [The Source](https://source.redhat.com/) (the first two that I looked at) provide a "sign out" action, so presumably we should, too.

> the right course of action would be to redirect the user back to the Auth page and remove all the cookies that the dashboard is using.

I think there are two interconnected things here:  there is "logging out" of the Dashboard (to wit, removing cookies, etc.) and there is ending the Dashboard's SSO session.  We should probably have the first of those encapsulated in a function so that we can call it from the various places that we need it (this one above being the second).  Whether the function also ends the SSO session remains TBD.

Reply:
Yeah, I did some tests, and I believe each user has a separate session for each client. Logging out of one client does not affect the logout of another client. 

> As for session _expiration_, I was assuming that that would, roughly speaking, never happen, because the Keycloak object in the Dashboard would keep renewing the token (and, thereby, I boldly assume, the session expiration). 

That is correct, as long as the dashboard tab is not idle for a duration that is configured in the Keycloak. This is represented by a refresh token expiration. If the user interacts with the dashboard before the refresh token is expired (currently 30 mins) then the refresh token also gets "refreshed" and the idle session has a new expiry.

> (Am I conflating two independent things, here?) I suppose an idle session in an open browser (perhaps in an inactive tab) would _eventually_ expire, at which point we _should_ kick the user back to the login page (assuming that the Keycloak object doesn't already handle this, too).

Keycloak wouldn't automatically redirect the user to auth page but if the session is expired and the user is to access the `ProtectedRoute` then our `ProtectedRoute` definition in the `App.js` would kick the user to the `auth` page for login. Inside the `ProtectedRoute` I am checking the custom `isLoggedIn` cookie which has an expiry same as the refresh token expiry so if our idle session is expired our cookie will also expire.


Now the issue here with logout is that with every reload of the `overview` page, it just logs the user out and that's really not a good user experience. I'll create an issue to work on it next but until we have that fix in, it may not be an ideal solution to ask users to logout on each overview reload. Instead as an intermediate fix, we could just redirect the user back to the Auth page and clear all the cookies. If the user indeed has an active SSO session then the user will be logged in (without having to re-enter the credentials) and redirected back to the home page (managed by the keycloak object).

Reply:
> Now the issue here with logout is that with every reload of the `overview` page, it just logs the user out and that's really not a good user experience. I'll create an issue to work on it next but until we have that fix in, it may not be an ideal solution to ask users to logout on each overview reload.

I'm not sure I understand all the implicit contexts here. Are you really saying that the existing code logs the user out, and that's not something that can be addressed in this PR? Or are you suggesting that somehow adding the *ability* to log out would cause that to happen? Or something else entirely?

Being able to "sign out" is probably a good thing, at which point they should be able to browse anonymously, or log in again (perhaps as a different user; e.g., to become `admin` for management). But we should never (even on page reload, and ideally across a browser restart) implicitly log out a user or be unable to implicitly continue a session. (Unless the OIDC token refresh fails due to external problems.)

Reply:
> I'm not sure I understand all the implicit contexts here.

Yeah, I'm having the same problem.  🙂 

> If the user interacts with the dashboard before the refresh token is expired (currently 30 mins) then the refresh token also gets "refreshed" and the idle session has a new expiry.

What exactly is a "session"?  Is there a separate "session token" or similar which is included in the HTTP requests?  Or, is this token just the refresh token?  (When we make a request for a new access token using the refresh token, do we get back both a new access token and a new refresh token, each with updated expirations?)

How do we know when we've been idle too long or when the session has expired?  In the absence of the Keycloak plugin, I would expect that the Dashboard would send a request to the Pbench Server, be told that our access token had expired, make a refresh request to Keycloak, and be told that the refresh token had expired.  Now that we're using the Keycloak plug-in, it is magically doing all of the token handling (including attaching it to the request to the Pbench Server), right?  And, this includes refreshing the access token when it expires.  So, how does the Dashboard code find out when the refresh has failed (for whatever reason)?

> Keycloak wouldn't automatically redirect the user to auth page but if the session is expired and the user is to access the `ProtectedRoute` then our `ProtectedRoute` definition in the `App.js` would kick the user to the auth page for login.

I'm unaware of the details of how the Keycloak plugin interacts with the Dashboard routing and with its requests to the Pbench Server, but given that the `ReactKeycloakProvider` is wrapped around our `BrowserRouter`, it seems like it could do things like redirect the user to the OIDC login page if the user makes a request after the session has expired.

> Inside the `ProtectedRoute` I am checking the custom `isLoggedIn` cookie which has an expiry same as the refresh token expiry so if our idle session is expired our cookie will also expire.

Yeah, and this is makes me suspicious:  the fact that feel that we have to maintain a cookie with the session expiration time so that we can tell whether the user is still logged in smells funny to me.  For the most part, it seems to me that the Dashboard code shouldn't care:  the login-ed-ness is all handled by the Keycloak plug-in (I think!) -- all that the Dashboard has to do is make requests of the Pbench Server and Keycloak will make sure that the tokens are there...right?  We might want to know whether the user has never logged in, so that we can support the anonymous user experience (or the initial, pre-logged-in experience), but I want to believe that we don't need to handle the session expiration (I would expect the Keycloak plug-in to deal with that).

> Now the issue here with logout is that with every reload of the overview page, it just logs the user out and that's really not a good user experience.

I assert that, if the user feels impelled to hit the browser reload button, then we're already deep into "not good user experience", and an additional rough edge (like having the user see a logout and then an auto-log-in) is small potatoes (so long as the user doesn't end up back at the OIDC login page...but, even that wouldn't bother me too much -- after all, the user _did_ hit the browser's reload button, so where did s/he think the experience would restart from?).

> Instead as an intermediate fix, we could just redirect the user back to the Auth page and clear all the cookies.

This sounds suspiciously close to what "logout" does.  The only difference is whether we reach out to Keycloak to request a logout from there, which I agree we do not want to do unless the user directs us to.

What I'm requesting is that we have common code (a single function) which encapsulates the _Dashboard_ end-session or logout or whatever want to term it, and that we use it for end-session and logout (and, if appropriate, logout can take the additional step of sending a logout request to the Keycloak server).

> we should never (even on page reload, and ideally across a browser restart) implicitly log out a user or be unable to implicitly continue a session.

I disagree.  From a security perspective, there really should be an external way to terminate a session, such that further access is denied without presenting credentials again.  Typically, closing the tab (or window) should suffice; restarting the browser should _definitely_ result in subsequently requiring a new login.  I don't know where reloading the page should fall in there...it would probably be a better experience if the login survived that...but it shouldn't survive restarting the browser!

Reply:
> I disagree. From a security perspective, there really should be an external way to terminate a session, such that further access is denied without presenting credentials again. Typically, closing the tab (or window) should suffice; restarting the browser should _definitely_ result in subsequently requiring a new login. I don't know where reloading the page should fall in there...it would probably be a better experience if the login survived that...but it shouldn't survive restarting the browser!

There definitely is an __external__ way to "terminate a session" through Keycloak and the SSO provider. On the next refresh cycle (which is short) the dashboard's token would be rejected. We can also have a local "logout" ... that's fine.

Browser restart? Maybe; that's what the "keep me logged in" checkbox on a lot of web sites (and until now on ours, albeit completely disconnected from any logic) is for. This should be orthogonal to how we track login, but has to do with tracking local database or cookie lifetime.

Ideally, in any web app framework, refreshing the active page should continue exactly where you were. There are all sorts of reasons, from display or network glitches to just wanting fresh data. If __this__ PR goes in with a messy experience on reload, well, fine; we need to get Keycloak up and running... but we shouldn't accept that as OK, and there's no reason for it as it's got nothing to do with the lifetime or management cycle of the auth and refresh tokens for Keycloak.

Reply:
> There definitely is an external way to "terminate a session" through Keycloak and the SSO provider.

Yeah, if the user is clever enough to go to some _other_ web site to log out of ours.  I think it's a big ask to expect the user to know s/he has that option.  (I'm not sure I know how to do that with RH SSO, e.g.)  What I meant by "external" was something over which the user has immediate control, like closing the tab/window or shutting down the browser.

However, you're absolutely right, there's that "keep me logged in" gizmo which is exactly intended to survive my "external" controls.

> Ideally, in any web app framework, refreshing the active page should continue exactly where you were.

"Exactly"...minus any error state or other accumulated cruft and with the addition of new data...so, perhaps not so much "exactly".  😆

I'm certainly not saying that we shouldn't support having our app reloaded.  However, I would prefer that we make it so that it doesn't occur to the user that doing so might be desirable.  As for this PR, I'm not worried about what happens on reload, so long as the result isn't a state worse than what it was before reloading.  And, yes, it would be lovely if app-reload and Keycloak integration were orthogonal, but I expect that it frequently won't be the case.

Reply:
> I'm not sure I understand all the implicit contexts here. Are you really saying that the existing code logs the user out, and that's not something that can be addressed in this PR? Or are you suggesting that somehow adding the _ability_ to log out would cause that to happen? Or something else entirely?

Existing code meaning the one I added in this PR? If yes then yeah what I was suggesting is that implicitly calling `keycloak.logout()` will forcefully log the user out of its current session. If we don't call the `keycloak.logout()` implicitly and only do the local cleanups of the cookies and redirect to the auth page then the forceful logout would not happen.
 

> What exactly is a "session"? Is there a separate "session token" or similar which is included in the HTTP requests? Or, is this token just the refresh token? (When we make a request for a new access token using the refresh token, do we get back both a new access token and a new refresh token, each with updated expirations?)

Yes, we get back a new access token and a new refresh token, each with updated expirations.

> How do we know when we've been idle too long or when the session has expired?

The dashboard can keep track of it internally by checking the `isLoggedIn` cookie and seeing if its already expired 

> In the absence of the Keycloak plugin, I would expect that the Dashboard would send a request to the Pbench Server, be told that our access token had expired, make a refresh request to Keycloak, and be told that the refresh token had expired. Now that we're using the Keycloak plug-in, it is magically doing all of the token handling (including attaching it to the request to the Pbench Server), right?

Keycloak plugin doesn't attach the access token with our API request automatically, we explicitly attach it in the `axiosInstance.js` file using interceptors. `req.headers.Authorization = Bearer ${keycloak?.token}` and we only do that if the `keycloak.authenticated ` returns `true`. So ideally we will never make an API request with an expired token and if the user is authenticated then it means the `keycloak.token` is valid (The refreshing part is managed by the Keycloak-js plugin). 

> Yeah, and this is makes me suspicious: the fact that feel that we have to maintain a cookie with the session expiration time so that we can tell whether the user is still logged in smells funny to me.

Right now the need for the cookie is because we reinitialize the keycloak object upon each page reload. 

> For the most part, it seems to me that the Dashboard code shouldn't care: the login-ed-ness is all handled by the Keycloak plug-in (I think!) -- all that the Dashboard has to do is make requests of the Pbench Server and Keycloak will make sure that the tokens are there...right? We might want to know whether the user has never logged in, so that we can support the anonymous user experience (or the initial, pre-logged-in experience), but I want to believe that we don't need to handle the session expiration (I would expect the Keycloak plug-in to deal with that).

Right but the dashboard would also need the ability to wait for the keycloak to check the current SSO of the user and when the keycloak report its status back then make a decision. Right now, the code I added in the PR does not have that capability.